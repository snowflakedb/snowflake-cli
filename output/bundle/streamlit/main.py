/Users/mraba/sources/snowflake-cli.worktree/SNOW-1980772-helper-for-displaying-snowsql-set-env-vars/main.py