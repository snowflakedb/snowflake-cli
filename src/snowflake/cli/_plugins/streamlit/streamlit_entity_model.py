# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import annotations

from typing import Literal, Optional

from pydantic import Field
from snowflake.cli.api.project.schemas.entities.common import (
    Artifacts,
    EntityModelBaseWithArtifacts,
    ExternalAccessBaseModel,
    GrantBaseModel,
    ImportsBaseModel,
)
from snowflake.cli.api.project.schemas.updatable_model import (
    DiscriminatorField,
)


class StreamlitEntityModel(
    EntityModelBaseWithArtifacts,
    ExternalAccessBaseModel,
    ImportsBaseModel,
    GrantBaseModel,
):
    type: Literal["streamlit"] = DiscriminatorField()  # noqa: A003
    title: Optional[str] = Field(
        title="Human-readable title for the Streamlit dashboard", default=None
    )
    comment: Optional[str] = Field(title="Comment for the Streamlit app", default=None)
    query_warehouse: str = Field(
        title="Snowflake warehouse to host the app", default=None
    )
    main_file: Optional[str] = Field(
        title="Entrypoint file of the Streamlit app", default="streamlit_app.py"
    )
    pages_dir: Optional[str] = Field(title="Streamlit pages", default=None)
    stage: Optional[str] = Field(
        title="Stage in which the appâ€™s artifacts will be stored", default="streamlit"
    )
    # Artifacts were optional, so to avoid BCR, we need to make them optional here as well
    artifacts: Optional[Artifacts] = Field(
        title="List of paths or file source/destination pairs to add to the deploy root",
        default=None,
    )
