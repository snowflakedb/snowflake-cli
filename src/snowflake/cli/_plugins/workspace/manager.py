from functools import cached_property
from pathlib import Path
from typing import Dict

from snowflake.cli._plugins.workspace.context import ActionContext, WorkspaceContext
from snowflake.cli.api.cli_global_context import get_cli_context
from snowflake.cli.api.console import cli_console as cc
from snowflake.cli.api.entities.utils import EntityActions, get_sql_executor
from snowflake.cli.api.exceptions import InvalidProjectDefinitionVersionError
from snowflake.cli.api.project.definition import default_role
from snowflake.cli.api.project.schemas.entities.entities import (
    Entity,
    v2_entity_model_to_entity_map,
)
from snowflake.cli.api.project.schemas.project_definition import (
    DefinitionV20,
    ProjectDefinition,
)
from snowflake.cli.api.project.util import to_identifier


class WorkspaceManager:
    """
    Instantiates entity instances from entity models, providing higher-order functionality on entity compositions.
    """

    def __init__(self, project_definition: ProjectDefinition, project_root: Path):
        if not project_definition.meets_version_requirement("2"):
            raise InvalidProjectDefinitionVersionError(
                "2.x", project_definition.definition_version
            )
        self._entities_cache: Dict[str, Entity] = {}
        self._project_definition: DefinitionV20 = project_definition
        self._project_root = project_root

    def get_entity(self, entity_id: str):
        """
        Returns an entity instance with the given ID. If exists, reuses the previously returned instance, or instantiates a new one otherwise.
        """
        if entity_id in self._entities_cache:
            return self._entities_cache[entity_id]
        entity_model = self._project_definition.entities.get(entity_id, None)
        if entity_model is None:
            raise ValueError(f"No such entity ID: {entity_id}")
        entity_model_cls = entity_model.__class__
        entity_cls = v2_entity_model_to_entity_map[entity_model_cls]
        workspace_ctx = WorkspaceContext(
            console=cc,
            project_root=self.project_root,
            get_default_role=_get_default_role,
            get_default_warehouse=_get_default_warehouse,
        )
        self._entities_cache[entity_id] = entity_cls(entity_model, workspace_ctx)
        return self._entities_cache[entity_id]

    def perform_action(self, entity_id: str, action: EntityActions, *args, **kwargs):
        """
        Instantiates an entity of the given ID and calls the given action on it.
        """
        entity = self.get_entity(entity_id)
        if entity.supports(action):
            return entity.perform(action, self.action_ctx, *args, **kwargs)
        else:
            raise ValueError(f'This entity type does not support "{action.value}"')

    @property
    def project_root(self) -> Path:
        return self._project_root

    @cached_property
    def action_ctx(self) -> ActionContext:
        return ActionContext(
            get_entity=self.get_entity,
        )


def _get_default_role() -> str:
    role = default_role()
    if role is None:
        role = get_sql_executor().current_role()
    return role


def _get_default_warehouse() -> str | None:
    warehouse = get_cli_context().connection.warehouse
    if warehouse:
        warehouse = to_identifier(warehouse)
    return warehouse
