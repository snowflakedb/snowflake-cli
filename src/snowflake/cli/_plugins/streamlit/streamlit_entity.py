from pathlib import Path
from typing import Optional

from click import ClickException
from snowflake.cli._plugins.connection.util import make_snowsight_url
from snowflake.cli._plugins.nativeapp.artifacts import build_bundle
from snowflake.cli._plugins.nativeapp.feature_flags import FeatureFlag
from snowflake.cli._plugins.streamlit.streamlit_entity_model import (
    StreamlitEntityModel,
)
from snowflake.cli._plugins.workspace.context import ActionContext
from snowflake.cli.api.entities.common import EntityBase
from snowflake.cli.api.project.project_paths import bundle_root
from snowflake.cli.api.project.schemas.entities.common import PathMapping
from snowflake.connector.cursor import SnowflakeCursor


class StreamlitEntity(EntityBase[StreamlitEntityModel]):
    """
    A Streamlit app.
    """

    def __init__(self, *args, **kwargs):
        if not FeatureFlag.ENABLE_NATIVE_APP_CHILDREN.is_enabled():
            raise NotImplementedError("Streamlit entity is not implemented yet")
        super().__init__(*args, **kwargs)

    @property
    def root(self):
        return self._workspace_ctx.project_root

    @property
    def artifacts(self):
        return self._entity_model.artifacts

    def action_bundle(self, action_ctx: ActionContext, *args, **kwargs):
        return self.bundle()

    def action_deploy(self, action_ctx: ActionContext, *args, **kwargs):
        # After adding bundle map- we should use it's mapping here
        # To copy artifacts to destination on stage.

        return self.deploy()

    def action_drop(self, action_ctx: ActionContext, *args, **kwargs):
        return self._execute_query(self.get_drop_sql())

    def action_execute(
        self, action_ctx: ActionContext, *args, **kwargs
    ) -> SnowflakeCursor:
        return self._execute_query(self.get_execute_sql())

    def action_get_url(
        self, action_ctx: ActionContext, *args, **kwargs
    ):  # maybe this should be a property
        name = self._entity_model.fqn.using_connection(self._conn)
        return make_snowsight_url(
            self._conn, f"/#/streamlit-apps/{name.url_identifier}"
        )

    def bundle(self, output_dir: Optional[Path] = None):
        build_bundle(
            self.root,
            output_dir or bundle_root(self.root, "streamlit"),
            [
                PathMapping(
                    src=artifact.src, dest=artifact.dest, processors=artifact.processors
                )
                for artifact in self._entity_model.artifacts
            ],
        )

    def deploy(self, *args, **kwargs):
        return self._execute_query(self.get_deploy_sql())

    def action_share(
        self, action_ctx: ActionContext, to_role: str, *args, **kwargs
    ) -> SnowflakeCursor:
        return self._execute_query(self.get_share_sql(to_role))

    def get_deploy_sql(
        self,
        if_not_exists: bool = False,
        replace: bool = False,
        from_stage_name: Optional[str] = None,
        artifacts_dir: Optional[Path] = None,
        schema: Optional[str] = None,
        *args,
        **kwargs,
    ):
        if replace and if_not_exists:
            raise ClickException("Cannot specify both replace and if_not_exists")

        if replace:
            query = "CREATE OR REPLACE "
        elif if_not_exists:
            query = "CREATE IF NOT EXISTS "
        else:
            query = "CREATE "

        schema_to_use = schema or self._entity_model.fqn.schema
        query += f"STREAMLIT {self._entity_model.fqn.set_schema(schema_to_use).sql_identifier}"

        if from_stage_name:
            query += f"\nROOT_LOCATION = '{from_stage_name}'"
        elif artifacts_dir:
            query += f"\nFROM '{artifacts_dir}'"

        query += f"\nMAIN_FILE = '{self._entity_model.main_file}'"

        if self.model.imports:
            query += "\n" + self.model.get_imports_sql()

        if self.model.query_warehouse:
            query += f"\nQUERY_WAREHOUSE = '{self.model.query_warehouse}'"

        if self.model.title:
            query += f"\nTITLE = '{self.model.title}'"

        if self.model.comment:
            query += f"\nCOMMENT = '{self.model.comment}'"

        if self.model.external_access_integrations:
            query += "\n" + self.model.get_external_access_integrations_sql()

        if self.model.secrets:
            query += "\n" + self.model.get_secrets_sql()

        return query + ";"

    def get_share_sql(self, to_role: str) -> str:
        return f"GRANT USAGE ON STREAMLIT {self.model.fqn.sql_identifier} TO ROLE {to_role};"

    def get_execute_sql(self):
        return f"EXECUTE STREAMLIT {self._entity_model.fqn}();"

    def get_usage_grant_sql(self, app_role: str, schema: Optional[str] = None) -> str:
        entity_id = self.entity_id
        streamlit_name = f"{schema}.{entity_id}" if schema else entity_id
        return (
            f"GRANT USAGE ON STREAMLIT {streamlit_name} TO APPLICATION ROLE {app_role};"
        )
