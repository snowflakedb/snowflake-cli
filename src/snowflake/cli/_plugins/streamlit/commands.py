# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import logging
from pathlib import Path
from typing import Dict

import click
import typer
from click import ClickException, UsageError
from snowflake.cli._plugins.object.command_aliases import (
    add_object_command_aliases,
    scope_option,
)
from snowflake.cli._plugins.streamlit.manager import StreamlitManager
from snowflake.cli._plugins.streamlit.streamlit_entity_model import (
    StreamlitEntityModel,
)
from snowflake.cli._plugins.streamlit.streamlit_project_paths import (
    StreamlitProjectPaths,
)
from snowflake.cli.api.cli_global_context import get_cli_context
from snowflake.cli.api.commands.decorators import (
    with_experimental_behaviour,
    with_project_definition,
)
from snowflake.cli.api.commands.flags import (
    ReplaceOption,
    entity_argument,
    identifier_argument,
    like_option,
)
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.constants import ObjectType
from snowflake.cli.api.exceptions import NoProjectDefinitionError
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.output.types import (
    CommandResult,
    MessageResult,
    SingleQueryResult,
)
from snowflake.cli.api.project.definition_conversion import (
    convert_project_definition_to_v2,
)

app = SnowTyperFactory(
    name="streamlit",
    help="Manages a Streamlit app in Snowflake.",
)
log = logging.getLogger(__name__)

StreamlitNameArgument = identifier_argument(
    sf_object="Streamlit app", example="my_streamlit"
)
OpenOption = typer.Option(
    False,
    "--open",
    help="Whether to open the Streamlit app in a browser.",
    is_flag=True,
)


add_object_command_aliases(
    app=app,
    object_type=ObjectType.STREAMLIT,
    name_argument=StreamlitNameArgument,
    like_option=like_option(
        help_example='`list --like "my%"` lists all streamlit apps that begin with “my”'
    ),
    scope_option=scope_option(help_example="`list --in database my_db`"),
)


@app.command(requires_connection=True)
def execute(
    name: FQN = StreamlitNameArgument,
    **options,
):
    """
    Executes a streamlit in a headless mode.
    """
    _ = StreamlitManager().execute(app_name=name)
    return MessageResult(f"Streamlit {name} executed.")


@app.command("share", requires_connection=True)
def streamlit_share(
    name: FQN = StreamlitNameArgument,
    to_role: str = typer.Argument(
        ...,
        help="Role with which to share the Streamlit app.",
        show_default=False,
    ),
    **options,
) -> CommandResult:
    """
    Shares a Streamlit app with another role.
    """
    cursor = StreamlitManager().share(streamlit_name=name, to_role=to_role)
    return SingleQueryResult(cursor)


def _default_file_callback(param_name: str):
    from click.core import ParameterSource  # type: ignore

    def _check_file_exists_if_not_default(ctx: click.Context, value):
        if (
            ctx.get_parameter_source(param_name) != ParameterSource.DEFAULT  # type: ignore
            and value
            and not Path(value).exists()
        ):
            raise ClickException(f"Provided file {value} does not exist")
        return Path(value)

    return _check_file_exists_if_not_default


@app.command("deploy", requires_connection=True)
@with_project_definition()
@with_experimental_behaviour()
def streamlit_deploy(
    replace: bool = ReplaceOption(
        help="Replace the Streamlit app if it already exists."
    ),
    entity_id: str = entity_argument("streamlit"),
    open_: bool = OpenOption,
    **options,
) -> CommandResult:
    """
    Deploys a Streamlit app defined in the project definition file (snowflake.yml). By default, the command uploads
    environment.yml and any other pages or folders, if present. If you don’t specify a stage name, the `streamlit`
    stage is used. If the specified stage does not exist, the command creates it. If multiple Streamlits are defined
    in snowflake.yml and no entity_id is provided then command will raise an error.
    """

    cli_context = get_cli_context()
    pd = cli_context.project_definition
    if not pd.meets_version_requirement("2"):
        if not pd.streamlit:
            raise NoProjectDefinitionError(
                project_type="streamlit", project_root=cli_context.project_root
            )
        pd = convert_project_definition_to_v2(cli_context.project_root, pd)

    streamlits: Dict[str, StreamlitEntityModel] = pd.get_entities_by_type(
        entity_type="streamlit"
    )

    streamlit_project_paths = StreamlitProjectPaths(cli_context.project_root)

    if not streamlits:
        raise NoProjectDefinitionError(
            project_type="streamlit", project_root=cli_context.project_root
        )

    if entity_id and entity_id not in streamlits:
        raise UsageError(f"No '{entity_id}' entity in project definition file.")

    if len(streamlits.keys()) == 1:
        entity_id = list(streamlits.keys())[0]

    if entity_id is None:
        raise UsageError(
            "Multiple Streamlit apps found. Please provide entity id for the operation."
        )

    # Get first streamlit
    streamlit: StreamlitEntityModel = streamlits[entity_id]
    url = StreamlitManager().deploy(
        streamlit=streamlit,
        streamlit_project_paths=streamlit_project_paths,
        replace=replace,
    )

    if open_:
        typer.launch(url)

    return MessageResult(f"Streamlit successfully deployed and available under {url}")


@app.command("get-url", requires_connection=True)
def get_url(
    name: FQN = StreamlitNameArgument,
    open_: bool = OpenOption,
    **options,
):
    """Returns a URL to the specified Streamlit app"""
    url = StreamlitManager().get_url(streamlit_name=name)
    if open_:
        typer.launch(url)
    return MessageResult(url)
