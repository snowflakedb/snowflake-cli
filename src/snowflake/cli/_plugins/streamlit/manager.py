# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import logging
from typing import List, Optional

from click import ClickException
from snowflake.cli._plugins.connection.util import (
    MissingConnectionAccountError,
    MissingConnectionRegionError,
    make_snowsight_url,
)
from snowflake.cli._plugins.object.manager import ObjectManager
from snowflake.cli._plugins.stage.manager import StageManager
from snowflake.cli._plugins.streamlit.streamlit_entity_model import (
    StreamlitEntityModel,
)
from snowflake.cli._plugins.streamlit.streamlit_project_paths import (
    StreamlitProjectPaths,
)
from snowflake.cli.api.artifacts.upload import put_files
from snowflake.cli.api.commands.experimental_behaviour import (
    experimental_behaviour_enabled,
)
from snowflake.cli.api.console import cli_console
from snowflake.cli.api.feature_flags import FeatureFlag
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.project.schemas.entities.common import PathMapping
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.connector.cursor import SnowflakeCursor
from snowflake.connector.errors import ProgrammingError

log = logging.getLogger(__name__)


class StreamlitManager(SqlExecutionMixin):
    def execute(self, app_name: FQN):
        query = f"EXECUTE STREAMLIT {app_name.sql_identifier}()"
        return self.execute_query(query=query)

    def share(self, streamlit_name: FQN, to_role: str) -> SnowflakeCursor:
        return self.execute_query(
            f"grant usage on streamlit {streamlit_name.sql_identifier} to role {to_role}"
        )

    def _put_streamlit_files(
        self,
        streamlit_project_paths: StreamlitProjectPaths,
        stage_root: str,
        artifacts: Optional[List[PathMapping]] = None,
    ):
        cli_console.step(f"Deploying files to {stage_root}")
        put_files(
            project_paths=streamlit_project_paths,
            stage_root=stage_root,
            artifacts=artifacts,
        )

    def _create_streamlit(
        self,
        streamlit: StreamlitEntityModel,
        replace: Optional[bool] = None,
        experimental: Optional[bool] = None,
        from_stage_name: Optional[str] = None,
    ):
        streamlit_id = streamlit.fqn.using_connection(self._conn)
        cli_console.step(f"Creating {streamlit_id} Streamlit")
        query = []
        if replace:
            query.append(f"CREATE OR REPLACE STREAMLIT {streamlit_id.sql_identifier}")
        elif experimental:
            # For experimental behaviour, we need to use CREATE STREAMLIT IF NOT EXISTS
            # for a streamlit app with an embedded stage
            # because this is analogous to the behavior for non-experimental
            # deploy which does CREATE STAGE IF NOT EXISTS
            query.append(
                f"CREATE STREAMLIT IF NOT EXISTS {streamlit_id.sql_identifier}"
            )
        else:
            query.append(f"CREATE STREAMLIT {streamlit_id.sql_identifier}")

        if from_stage_name:
            query.append(f"ROOT_LOCATION = '{from_stage_name}'")

        query.append(f"MAIN_FILE = '{streamlit.main_file}'")
        if streamlit.imports:
            query.append(streamlit.get_imports_sql())

        if not streamlit.query_warehouse:
            cli_console.warning(
                "[Deprecation] In next major version we will remove default query_warehouse='streamlit'."
            )
            query.append(f"QUERY_WAREHOUSE = 'streamlit'")
        else:
            query.append(f"QUERY_WAREHOUSE = {streamlit.query_warehouse}")

        if streamlit.title:
            query.append(f"TITLE = '{streamlit.title}'")

        if streamlit.comment:
            query.append(f"COMMENT = '{streamlit.comment}'")

        if streamlit.external_access_integrations:
            query.append(streamlit.get_external_access_integrations_sql())

        if streamlit.secrets:
            query.append(streamlit.get_secrets_sql())

        self.execute_query("\n".join(query))

    def deploy(
        self,
        streamlit: StreamlitEntityModel,
        streamlit_project_paths: StreamlitProjectPaths,
        replace: bool = False,
    ):
        streamlit_id = streamlit.fqn.using_connection(self._conn)
        if (
            ObjectManager().object_exists(object_type="streamlit", fqn=streamlit_id)
            and not replace
        ):
            raise ClickException(
                f"Streamlit {streamlit.fqn} already exist. If you want to replace it use --replace flag."
            )

        # for backwards compatibility - quoted stage path might be case-sensitive
        # https://docs.snowflake.com/en/sql-reference/identifiers-syntax#double-quoted-identifiers
        streamlit_name_for_root_location = streamlit_id.name
        use_versioned_stage = FeatureFlag.ENABLE_STREAMLIT_VERSIONED_STAGE.is_enabled()
        if (
            experimental_behaviour_enabled()
            or FeatureFlag.ENABLE_STREAMLIT_EMBEDDED_STAGE.is_enabled()
            or use_versioned_stage
        ):
            """
            1. Create streamlit object
            2. Upload files to embedded stage
            """
            # TODO: Support from_stage
            # from_stage_stmt = f"FROM_STAGE = '{stage_name}'" if stage_name else ""
            self._create_streamlit(
                streamlit=streamlit,
                replace=replace,
                experimental=True,
            )
            try:
                if use_versioned_stage:
                    self.execute_query(
                        f"ALTER STREAMLIT {streamlit_id.identifier} ADD LIVE VERSION FROM LAST"
                    )
                elif not FeatureFlag.ENABLE_STREAMLIT_NO_CHECKOUTS.is_enabled():
                    self.execute_query(
                        f"ALTER streamlit {streamlit_id.identifier} CHECKOUT"
                    )
            except ProgrammingError as e:
                # If an error is raised because a CHECKOUT has already occurred or a LIVE VERSION already exists, simply skip it and continue
                if "Checkout already exists" in str(
                    e
                ) or "There is already a live version" in str(e):
                    log.info("Checkout already exists, continuing")
                else:
                    raise

            stage_path = streamlit_id.identifier
            embedded_stage_name = f"snow://streamlit/{stage_path}"
            if use_versioned_stage:
                # "LIVE" is the only supported version for now, but this may change later.
                stage_root = f"{embedded_stage_name}/versions/live"
            else:
                stage_root = f"{embedded_stage_name}/default_checkout"

            self._put_streamlit_files(
                streamlit_project_paths,
                stage_root,
                streamlit.artifacts,
            )
        else:
            """
            1. Create stage
            2. Upload files to created stage
            3. Create streamlit from stage
            """
            stage_manager = StageManager()

            stage_name = streamlit.stage or "streamlit"
            stage_name = FQN.from_string(stage_name).using_connection(self._conn)

            cli_console.step(f"Creating {stage_name} stage")
            stage_manager.create(fqn=stage_name)

            stage_root = stage_manager.get_standard_stage_prefix(
                f"{stage_name}/{streamlit_name_for_root_location}"
            )

            self._put_streamlit_files(
                streamlit_project_paths, stage_root, streamlit.artifacts
            )

            self._create_streamlit(
                streamlit=streamlit,
                replace=replace,
                from_stage_name=stage_root,
                experimental=False,
            )

            self.grant_privileges(streamlit)

        return self.get_url(streamlit_name=streamlit_id)

    def grant_privileges(self, entity_model: StreamlitEntityModel):
        if not entity_model.grants:
            return
        for grant in entity_model.grants:
            self.execute_query(grant.get_grant_sql(entity_model))

    def get_url(self, streamlit_name: FQN) -> str:
        try:
            fqn = streamlit_name.using_connection(self._conn)
            return make_snowsight_url(
                self._conn,
                f"/#/streamlit-apps/{fqn.url_identifier}",
            )
        except (MissingConnectionRegionError, MissingConnectionAccountError) as e:
            return "https://app.snowflake.com"
