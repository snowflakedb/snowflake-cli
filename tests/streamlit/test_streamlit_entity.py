from __future__ import annotations

from pathlib import Path
from unittest import mock

import pytest
import yaml
from snowflake.cli._plugins.streamlit.streamlit_entity import (
    StreamlitEntity,
)
from snowflake.cli._plugins.streamlit.streamlit_entity_model import (
    StreamlitEntityModel,
)
from snowflake.cli._plugins.workspace.context import ActionContext, WorkspaceContext

from tests.testing_utils.mock_config import mock_config_key

STREAMLIT_NAME = "test_streamlit"
CONNECTOR = "snowflake.connector.connect"
CONTEXT = ""
EXECUTE_QUERY = "snowflake.cli.api.sql_execution.BaseSqlExecutor.execute_query"

GET_UI_PARAMETERS = "snowflake.cli._plugins.connection.util.get_ui_parameters"


@pytest.fixture
def example_streamlit_workspace(project_directory):
    with mock_config_key("enable_native_app_children", True):
        with project_directory("example_streamlit_v2") as pdir:
            with Path(pdir / "snowflake.yml").open() as definition_file:
                definition = yaml.safe_load(definition_file)
                model = StreamlitEntityModel(
                    **definition.get("entities", {}).get("test_streamlit")
                )

                workspace_context = WorkspaceContext(
                    console=mock.MagicMock(),
                    project_root=pdir,
                    get_default_role=lambda: "test_role",
                    get_default_warehouse=lambda: "test_warehouse",
                )

                yield (
                    StreamlitEntity(
                        workspace_ctx=workspace_context, entity_model=model
                    ),
                    ActionContext(
                        get_entity=lambda *args: None,
                    ),
                )


def test_cannot_instantiate_without_feature_flag():
    with pytest.raises(NotImplementedError) as err:
        StreamlitEntity()
    assert str(err.value) == "Streamlit entity is not implemented yet"


def test_nativeapp_children_interface(example_streamlit_workspace, snapshot):
    sl, action_context = example_streamlit_workspace

    sl.bundle()
    bundle_artifact = sl.root / "output" / "bundle" / "streamlit" / "streamlit_app.py"
    deploy_sql_str = sl.get_deploy_sql()
    grant_sql_str = sl.get_usage_grant_sql(app_role="app_role")

    assert bundle_artifact.exists()
    assert deploy_sql_str == snapshot
    assert (
        grant_sql_str == f"GRANT USAGE ON STREAMLIT None TO APPLICATION ROLE app_role;"
    )


def test_bundle(example_streamlit_workspace):
    entity, action_ctx = example_streamlit_workspace
    entity.action_bundle(action_ctx)

    output = entity.root / "output" / "bundle" / "streamlit"  # noqa

    assert output.exists()
    assert (output / "streamlit_app.py").exists()
    assert (output / "environment.yml").exists()
    assert (output / "pages" / "my_page.py").exists()


@mock.patch(EXECUTE_QUERY)
def test_deploy(mock_execute, example_streamlit_workspace):
    entity, action_ctx = example_streamlit_workspace
    entity.action_deploy(action_ctx)

    mock_execute.assert_called_with(
        f"CREATE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')\nMAIN_FILE = 'streamlit_app.py'\nQUERY_WAREHOUSE = 'test_warehouse'\nTITLE = 'My Fancy Streamlit';"
    )


@mock.patch(EXECUTE_QUERY)
def test_drop(mock_execute, example_streamlit_workspace):
    entity, action_ctx = example_streamlit_workspace
    entity.action_drop(action_ctx)

    mock_execute.assert_called_with(f"DROP STREAMLIT IDENTIFIER('{STREAMLIT_NAME}');")


@mock.patch(CONNECTOR)
@mock.patch(
    GET_UI_PARAMETERS,
    return_value={"UI_SNOWSIGHT_ENABLE_REGIONLESS_REDIRECT": "false"},
)
@mock.patch("click.get_current_context")
def test_get_url(
    mock_get_ctx,
    mock_param,
    mock_connect,
    mock_cursor,
    example_streamlit_workspace,
    mock_ctx,
):
    ctx = mock_ctx(
        mock_cursor(
            rows=[
                {"SYSTEM$GET_SNOWSIGHT_HOST()": "https://snowsight.domain"},
                {"SYSTEM$RETURN_CURRENT_ORG_NAME()": "FOOBARBAZ"},
                {"CURRENT_ACCOUNT_NAME()": "https://snowsight.domain"},
            ],
            columns=["SYSTEM$GET_SNOWSIGHT_HOST()"],
        )
    )
    mock_connect.return_value = ctx
    mock_get_ctx.return_value = ctx

    entity, action_ctx = example_streamlit_workspace
    result = entity.action_get_url(action_ctx)

    mock_connect.assert_called()


@mock.patch(EXECUTE_QUERY)
def test_share(mock_connect, example_streamlit_workspace):
    entity, action_ctx = example_streamlit_workspace
    entity.action_share(action_ctx, to_role="test_role")

    mock_connect.assert_called_with(
        "GRANT USAGE ON STREAMLIT IDENTIFIER('test_streamlit') TO ROLE test_role;"
    )


@mock.patch(EXECUTE_QUERY)
def test_execute(mock_execute, example_streamlit_workspace):
    entity, action_ctx = example_streamlit_workspace
    entity.action_execute(action_ctx)

    mock_execute.assert_called_with(f"EXECUTE STREAMLIT {STREAMLIT_NAME}();")


def test_get_execute_sql(example_streamlit_workspace):
    entity, action_ctx = example_streamlit_workspace
    execute_sql = entity.get_execute_sql()

    assert execute_sql == f"EXECUTE STREAMLIT {STREAMLIT_NAME}();"


def test_get_drop_sql(example_streamlit_workspace):
    entity, action_ctx = example_streamlit_workspace
    drop_sql = entity.get_drop_sql()

    assert drop_sql == f"DROP STREAMLIT IDENTIFIER('{STREAMLIT_NAME}');"


@pytest.mark.parametrize(
    "kwargs",
    [
        {"replace": True},
        {"if_not_exists": True},
        {"from_stage_name": "test_stage"},
        {"from_stage_name": "test_stage", "replace": True},
        {"from_stage_name": "test_stage", "if_not_exists": True},
    ],
)
def test_get_deploy_sql(kwargs, example_streamlit_workspace, snapshot):
    entity, action_ctx = example_streamlit_workspace
    deploy_sql = entity.get_deploy_sql(**kwargs)

    assert deploy_sql == snapshot
