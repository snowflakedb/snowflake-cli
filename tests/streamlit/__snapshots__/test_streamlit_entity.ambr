# serializer version: 1
# name: test_get_deploy_sql[kwargs0]
  '''
  CREATE OR REPLACE STREAMLIT IDENTIFIER('test_streamlit')
  MAIN_FILE = 'streamlit_app.py'
  QUERY_WAREHOUSE = 'test_warehouse'
  TITLE = 'My Fancy Streamlit';
  '''
# ---
# name: test_get_deploy_sql[kwargs1]
  '''
  CREATE IF NOT EXISTS STREAMLIT IDENTIFIER('test_streamlit')
  MAIN_FILE = 'streamlit_app.py'
  QUERY_WAREHOUSE = 'test_warehouse'
  TITLE = 'My Fancy Streamlit';
  '''
# ---
# name: test_get_deploy_sql[kwargs2]
  '''
  CREATE STREAMLIT IDENTIFIER('test_streamlit')
  ROOT_LOCATION = 'test_stage'
  MAIN_FILE = 'streamlit_app.py'
  QUERY_WAREHOUSE = 'test_warehouse'
  TITLE = 'My Fancy Streamlit';
  '''
# ---
# name: test_get_deploy_sql[kwargs3]
  '''
  CREATE OR REPLACE STREAMLIT IDENTIFIER('test_streamlit')
  ROOT_LOCATION = 'test_stage'
  MAIN_FILE = 'streamlit_app.py'
  QUERY_WAREHOUSE = 'test_warehouse'
  TITLE = 'My Fancy Streamlit';
  '''
# ---
# name: test_get_deploy_sql[kwargs4]
  '''
  CREATE IF NOT EXISTS STREAMLIT IDENTIFIER('test_streamlit')
  ROOT_LOCATION = 'test_stage'
  MAIN_FILE = 'streamlit_app.py'
  QUERY_WAREHOUSE = 'test_warehouse'
  TITLE = 'My Fancy Streamlit';
  '''
# ---
# name: test_nativeapp_children_interface
  '''
  CREATE STREAMLIT IDENTIFIER('test_streamlit')
  MAIN_FILE = 'streamlit_app.py'
  QUERY_WAREHOUSE = 'test_warehouse'
  TITLE = 'My Fancy Streamlit';
  '''
# ---
